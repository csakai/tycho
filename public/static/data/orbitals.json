[{
	"name": "Earth",
	"albedo": 0.367,
	"sidereal": 0.99726968,
	"radius": 6371,
	"markdown": "earth.md",
	"mass": 5.97237e+24,
	"centralMass": 1.989e+30,
	"axialTilt": 23.4392811,
	"inclination": 7.155,
	"atmosphere": 35260,
	"satellites": [{
		"name": "Moon",
		"albedo": 0.136,
		"sidereal": 0.99726968,
		"radius": 1737.1,
		"markdown": "moon.md",
		"mass": 5.97237e+24,
		"centralMass": 5.97237e+24,
		"axialTilt": 6.687,
		"inclination": 5.145,
		"atmosphere": 10066278,
		"ephemeris": {
			"nasaOrbitalCode": "301",
			"nasaBarycenterCode": "3",
			"revolutionOrder": "d"
		},
		"argumentOfPeriapsis": 282.5992409518417,
		"eccentricity": 0.03405532141489297,
		"longitudeOfAscendingNode": 142.3595722910625,
		"semimajor": 376126.5103019063,
		"semiminor": 375908.33786092995,
		"periapses": {
			"last": 1509927058416.626,
			"next": 1512377208118.1555
		},
		"id": "moon"
	}],
	"ephemeris": {
		"nasaOrbitalCode": "3",
		"nasaBarycenterCode": "0",
		"revolutionOrder": "y"
	},
	"argumentOfPeriapsis": 114.20763,
	"eccentricity": 0.01355360061112275,
	"longitudeOfAscendingNode": 348.73936,
	"semimajor": 150759239.9570005,
	"semiminor": 150745392.07806867,
	"periapses": {
		"last": 1136430000000,
		"next": 1167987600000
	},
	"id": "earth"
},{
	"name": "Venus",
	"albedo": 0.689,
	"sidereal": -243.025,
	"radius": 6051.8,
	"markdown": "venus.md",
	"mass": 4.8675e+24,
	"centralMass": 1.989e+30,
	"axialTilt": 2.64,
	"inclination": 3.86,
	"atmosphere": 16052641,
	"ephemeris": {
		"nasaOrbitalCode": "2",
		"nasaBarycenterCode": "0",
		"revolutionOrder": "y"
	},
	"argumentOfPeriapsis": 55.186,
	"eccentricity": 0.0067,
	"longitudeOfAscendingNode": 76.678,
	"semimajor": 108208179,
	"semiminor": 107477094,
	"periapses": {
		"last": 1429348560000,
		"next": 1448787300000
	},
	"id": "venus"
}]