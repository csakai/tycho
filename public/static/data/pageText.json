{"planetScale":"Planet Scale: × ","speedScale":"Speed Scale: × 10^","settings":"Settings"}
