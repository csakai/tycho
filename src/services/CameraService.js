import TWEEN from 'tween.js';
import {Vector3} from 'three';
import Constants from '../constants';
import OrbitalService from './OrbitalService';
import Scale from '../utils/Scale';

export default class CameraService {

  static CAMERA_INITIAL_POSITION = new Vector3(300, 300, 300) // TODO: const

  /**
   * Converts an instance of Vector3 to a plain object with x, y, z properties.
   *
   * @param {Vector3} v - vector to convert
   * @returns {Object} - plain object with x, y, z properties
   */
  static vectorToObject = (v) => {
    const {x, y, z} = v;
    return {x, y, z};
  }

  /**
   * Converts a plain object to an instance of Vector3.
   *
   * @param {Object} o - object, with x, y, z properties
   * @returns {Vector3} - converted vector
   */
  static objectToVector = (o) => {
    return new Vector3(o.x, o.y, o.z);
  }

  /**
   * Runs a given tween instance with the provided params.
   *
   * @param {Tween} tween - instance of tween
   * @param {Vector3} dest - destination vector
   * @param {Function} onComplete - callback
   */
  static startTween = (tween, dest, onComplete) => {
    return tween
      .easing(TWEEN.Easing.Quadratic.Out)
      .to(CameraService.vectorToObject(dest), Constants.WebGL.Tween.NORMAL)
      .onComplete(onComplete)
      .start();
  }

  /**
   * Calculates the min distance from target to camera to avoid collision.
   *
   * @param {Object[]} orbitals - list of orbitals
   * @param {String} targetName - id of active orbital target
   * @param {Number} scale - user-defined planet scale
   * @returns {Number} min distance
   */
  static getMinDistance = (orbitals, targetName, scale) => {
    const targetRadius = OrbitalService.getTargetByName(orbitals, targetName).radius;
    const scaledRadius = Scale(targetRadius, scale);

    return scaledRadius * 2;
  }
  
  /**
   * Creates an instance of Tween for moving a pivot to a given target.
   *
   * @param {Vector3} from - starting position
   * @param {Vector3} to - target position
   * @param {Object3D} target - new target
   * @param {Object3D} group - camera pivot
   * @returns {Tween}
   */
  static getPivotTween = (from, to, target, group) => {
    return new TWEEN
      .Tween(from)
      .to(to, 2000)
      .onUpdate(CameraService.setPivotPosition.bind(this, group, from))
      .onComplete(CameraService.movePivot.bind(this, group, target))
      .start();
  }

  /**
   * Moves the given pivot to the new target at <0>.
   *
   * @param {Object3D} group - camera pivot
   * @param {Object3D} target - new target
   */
  static movePivot = (group, target) => {
    target.add(group);
    group.position.set(0, 0, 0);
  }

  /**
   * Sets the camera pivot position.
   *
   * @param {Object3D} group - camera pivot
   * @param {Vector3} vect - new vector position
   */
  static setPivotPosition = (group, {x, y, z}) => {
    group.position.set(x, y, z);
  }

  /**
   * Returns the world position (i.e., w.r.t. <0>) of the given target.
   *
   * @param {Object3D} target - object to get world position of
   * @returns {Vector3} world position
   */
  static getWorldPosition = (target) => {
    target.updateMatrixWorld();

    const matrix = target.matrixWorld;
    const vect = new Vector3();

    vect.setFromMatrixPosition(matrix);

    return vect;
  }
}
