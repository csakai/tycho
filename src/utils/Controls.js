import * as THREE from 'three';
import OrbitControls from 'three-orbit-controls';
import TWEEN from 'tween.js';
import Constants from '../constants';

export default class Controls extends OrbitControls(THREE) {

  constructor(camera, domElement) {
    super(camera, domElement);

    this.camera = camera;
    this.enabled = true;
    this.enableZoom = false;
    this.enablePan = false;
    this.level = Constants.WebGL.Zoom.MAX;
    this.minDistance = Constants.WebGL.Camera.MIN_DISTANCE;
    this.maxDistance = Constants.WebGL.Camera.MAX_DISTANCE;
  }

  /**
   * Change zoom. Overrides the existing zoom tween in progress.
   *
   * @param {level} level - percentage of desired zoom level [0,100]
   */
  zoom = (level) => {
    if(this.level !== level) {
      this.pan(level / 100);
      this.level = level;
    }
  }

  /**
   * Gets the zoom delta from controls and invokes the given action if zoom has changed.
   *
   * @param {Event} ev - DOM wheel event
   * @param {Function} action - callback action
   */
  wheelZoom = (ev, action) => {
    const zoom = this.getZoomDelta(ev.deltaY);
    const current = Math.round(this.level * 100);

    if (current !== zoom) {
      action(zoom);
    }
  }

  /**
   * Calculates the zoom percentage from the given mousewheel delta.
   *
   * @param {Number} delta - mousewheel delta
   * @returns {Number} new zoom level
   */
  getZoomDelta = (delta) => {
    let zoom = this.level;
    
    zoom += (delta / this.getWheelDeltaDivisor(zoom));
    zoom = Math.max(zoom, 0);
    zoom = Math.min(zoom, 100);
   
    return zoom;
  }

  /**
   * Calculates a wheel delta that inflates based on how far away
   * the user is from the target, because outer space is really big.
   *
   * @param {Number} level - current zoom level 
   * @returns {Number} scrollwheel delta
   */
  getWheelDeltaDivisor = (level) => {
    let delta = Constants.UI.WHEEL_DELTA_DIVISOR;
    let match = false;

    Constants.WebGL.ScrollScale
      .forEach(({distance, scale}) => {
        if (level > distance && !match) {
          delta *= scale;
          match = true;
        }
      });

    return delta;
  }

  /**
   * Sets the current camera position to the scaled zoom vector.
   *
   * @param {Number} percent - percentage of zoom [0,1]
   */
  pan = (percent, log) => {
    let position = this.camera.position;
    let newVector = this.getZoomVector(position, this.maxDistance * percent);
    let minVector = this.getZoomVector(position, this.minDistance);
    
    if (newVector.length() >= this.minDistance) {
      position.copy(newVector);
    } else {
      position.copy(minVector);
    }
  }

  /**
   * Returns the current vector scaled to the desired magnitude.
   *
   * @param {Vector3} vector - current position vector
   * @param {Number} scalar - magnitude of vector
   * @returns {Vector3}
   */
  getZoomVector = (vector, scalar) => {
    return vector
      .clone()
      .normalize()
      .multiplyScalar(scalar);
  }

  /**
   * Returns the distance between <0> and current camera position.
   * 
   * @returns {Number} distance
   */
  getDistance = () => {
    return this.maxDistance - this.minDistance;
  }

  /**
   * Enable controls.
   */
  enable = () => {
    this.enabled = true;
  }

  /**
   * Disable controls.
   */
  disable = () => {
    this.enabled = false;
  }

  /**
   * Starts autorotating the scene clockwise.
   *
   * @param {Number} speed - speed to rotate at
   */
  startAutoRotate = (speed) => {
    this.autoRotate = true;
    this.autoRotateSpeed = speed;
  }

  /**
   * Stop autorotating the scene.
   */
  stopAutoRotate = () => {
    this.autoRotate = false;
  }

  /**
   * Deletes Tween instance and data.
   */
  endTween = () => {
    delete this.tweenData;
    delete this.tweenBase;
  }

  /**
   * Cancels Tween, if one is in progress.
   */
  cancelTween = () => {
    if (this.tweenBase) {
      this.tweenBase.stop();
      this.endTween();
    }
  }

  /**
   * Zooms to the level of the current Tween data.
   */
  updateTween = () => {
    this.zoom(this.tweenData.level);
  }

  /**
   * Invokes teardown methods for active tween.
   */
  completeTween = () => {
    if (this.tweenDone) {
      this.tweenDone(this.level);
    }
    this.endTween();
  }

  /**
   * Creates an animation to zoom to the specified level.
   *
   * @param {Number} level - percentage of zoom
   * @param {Function} onDone - animation completion callback
   */
  tweenZoom = (level, onDone) => {
    this.cancelTween();

    this.tweenDone = onDone;
    this.tweenData = {level: this.level};

    this.tweenBase = new TWEEN
      .Tween(this.tweenData)
      .easing(TWEEN.Easing.Quadratic.Out)
      .to({level}, Constants.WebGL.Tween.SLOW)
      .onUpdate(this.updateTween)
      .onComplete(this.completeTween)
      .start();
  }
}
