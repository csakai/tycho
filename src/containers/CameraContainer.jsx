import React from 'react';
import PropTypes from 'prop-types';
import CameraService from '../services/CameraService';
import Controls from '../utils/Controls';
import Constants from '../constants';

export default class CameraContainer extends React.Component {

  static propTypes = {
      ratio: PropTypes.number.isRequired,
      targetId: PropTypes.string,
      actions: PropTypes.object,
      speed: PropTypes.number,
      scale: PropTypes.number,
      zoom: PropTypes.number,
      scene: PropTypes.object,
      isAutoOrbitEnabled: PropTypes.bool,
      orbitalData: PropTypes.array
  }

  componentWillUnmount = () => {
      this.controls.dispose();
      delete this.controls;
  }

  componentWillReceiveProps = (nextProps) => {
      this.maybeCreateControls(nextProps);
      this.maybeMoveCameraPivot(nextProps);
      this.maybeUpdateControlsZoom(nextProps);
      this.maybeUpdateAutoOrbit(nextProps);
      this.maybePreventCameraCollision(nextProps);
  }

  /**
   * Updates controls, if mounted.
   */
  update = () => {
      if (this.controls) {
          this.controls.update();
      }
  }

  /**
   * Re-renders controls with updated canvas DOM Element.
   *
   * @param {Object} props
   * @param {HTMLElement} props.domElement
   */
  maybeCreateControls = ({domElement}) => {
      if (this.props.domElement !== domElement) {
          this.controls = new Controls(this.refs.camera, domElement);
      }
  }

  /**
   * Starts tweening the camera to the new target if targetId has changed.
   *
   * @param {Object} props
   * @param {String} props.targetId
   */
  maybeMoveCameraPivot = ({targetId}) => {
      if (this.props.targetId !== targetId) {
          this.movePivot(targetId, !!this.props.targetId);
      }
  }

  /**
   * Sets the min distance to the radius of the target, if the scale or target updated.
   * This prevents the camera from colliding with the target, should the zoom change.
   *
   * @param {Object} props
   * @param {String} props.targetId - id of active target
   * @param {Number} props.scale - user-defined planet scale
   */
  maybePreventCameraCollision = ({targetId, scale}) => {
      if (this.props.targetId !== targetId || this.props.scale !== scale) {
          this.controls.minDistance = CameraService
              .getMinDistance(this.props.orbitalData, targetId, scale);
      }
  }

  /**
   * Updates zoom level if the zoom prop has changed.
   *
   * @param {Object} props
   * @param {Number} props.zoom - new zoom prop value
   */
  maybeUpdateControlsZoom = ({zoom}) => {
      if (this.props.zoom !== zoom) {
          this.controls.zoom(zoom);
      }
  }

  /**
   * Updates the auto orbit state if the isAutoOrbitEnabled has changed.
   *
   * @param {Object} props
   * @param {Boolean} props.isAutoOrbitEnabled - new isAutoOrbitEnabled prop value
   */
  maybeUpdateAutoOrbit = ({isAutoOrbitEnabled}) => {
      if (this.props.isAutoOrbitEnabled !== isAutoOrbitEnabled) {
          this.controls.autoRotate = isAutoOrbitEnabled;
      }
  }

  /**
   * Starts Tween to the position having a key of targetId.
   *
   * @param {String} targetId - name of orbital to move to
   * @param {Boolean} animate - set to true if it should animate
   */
  movePivot = (targetId, animate) => {
      const {scene} = this.props;
      const {pivot} = this.refs;
      const target = scene.getObjectByName(targetId);

      if (target && animate) {
          this.setInteractivity(false);
          this.startTween(target, pivot, scene);
      }
  }

  /**
   * Cancels any Tween in progress.
   */
  cancelTween = () => {
      if (this.tweenBase) {
          this.tweenBase.stop();
          delete this.tweenBase;
      }
  }

  /**
   * Teardown for pivot tweening.
   */
  endTween = () => this.setInteractivity(true)

  /**
   * Starts a new Tween to the active target position.
   * Any Tween already in progress will be cancelled.
   *
   * @param {THREE.Object3D} target - target object
   * @param {THREE.Group} pivot - camera pivot
   */
  startTween = (target, pivot, scene) => {
      const v = CameraService.getWorldPosition(target);
      const w = CameraService.getWorldPosition(pivot);

      // set the pivot position to active position
      CameraService.attachToWorld(scene, pivot, w);

      this.cancelTween();
      this.zoomInFull();

      this.tweenBase = CameraService.getPivotTween(w, v, target, pivot, this.endTween);
  }

  /**
   * Enables/disables scene interactivity.
   *
   * @param {Boolean} disabled - whether or not controls are disabled
   */
  setInteractivity = (enabled) => {
      this.props.action.setUIControls(!!enabled);
      this.props.action.setPaused(enabled);
      this.controls.enabled = !!enabled;
  }

  /**
   * Tweens zoom to minimum allowable zoom.
   */
  zoomInFull = () => {
      this.controls.tweenZoom(Constants.WebGL.Zoom.MIN, this.props.action.changeZoom);
  }

  render() {
      return (
          <group ref="pivot">
              <perspectiveCamera
                  name="camera"
                  ref="camera"
                  aspect={this.props.ratio}
                  fov={Constants.WebGL.Camera.FOV}
                  near={Constants.WebGL.Camera.NEAR}
                  far={Constants.WebGL.Camera.FAR}
                  position={CameraService.CAMERA_INITIAL_POSITION}
              />
          </group>
      );
  }
}
