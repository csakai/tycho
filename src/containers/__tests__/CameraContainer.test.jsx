import React from 'react';
import React3 from 'react-three-renderer';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import TWEEN from 'tween.js';
import {Vector3, Camera, Object3D} from 'three';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import CameraContainer from '../CameraContainer';
import Controls from '../../utils/Controls';
import data from './__fixtures__/orbitals.json';
import CameraService from '../../services/CameraService';

jest.mock('../../services/CameraService');

const action = {
  setUIControls: jest.fn(),
  changeSpeed: jest.fn()
};

describe('Camera Container', () => {
  let component, cameraContainer;

  beforeEach(() => {
    component = shallow(
      <CameraContainer 
        action={action}
        targetName="testPlanet"
        orbitalData={data}
        ratio={1.5}
      />
    );
    cameraContainer = component.instance();
  });

  describe('Prop update methods', () => {
    beforeEach(() => {
      cameraContainer.controls = new Controls(new Camera());
    });

    describe('componentDidMount()', () => {
      it('should initialize a new instance of controls', () => {
        cameraContainer.refs = {camera: new Camera()};
        cameraContainer.componentDidMount();

        expect(cameraContainer).toHaveProperty('controls');
        expect(cameraContainer.controls).toBeInstanceOf(Controls);
      });
    });

    describe('componentWillUnmount()', () => {
      beforeEach(() => {
        const camera = new Camera();
        cameraContainer.controls = new Controls(camera);
      });

      it('should dispose of the controls', () => {
        const spy = jest.spyOn(cameraContainer.controls, 'dispose');

        cameraContainer.componentWillUnmount();

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledTimes(1);
      });

      it('should no longer have controls assigned to the class instance', () => {
        cameraContainer.componentWillUnmount();
        expect(cameraContainer).not.toHaveProperty('controls');
      });
    });

    describe('componentWillReceiveProps()', () => {
      const scene = {
        getObjectByName: () => new Object3D(),
        add: jest.fn()
      };

      it('should start tweening to `targetName` if the prop has changed', () => {
        const targetName = 'Mars';
        const nextProps = {targetName: 'Earth'};

        cameraContainer.startTween = jest.fn();
        cameraContainer.props = {targetName, action, scene};
        cameraContainer.componentWillReceiveProps(nextProps);

        const spy = jest.spyOn(cameraContainer, 'startTween');
        
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith(nextProps.targetName);
      });

      it('should not tween if the `targetName` if the prop has not changed', () => {
        const targetName = 'Mars';
        const nextProps = {targetName};
        const spy = jest.spyOn(cameraContainer, 'startTween');

        cameraContainer.props = {targetName, action, scene};
        cameraContainer.componentWillReceiveProps(nextProps);

        expect(spy).not.toHaveBeenCalled();
      });
    });

    describe('maybePreventCameraCollision()', () => {
      it('should update the minDistance if the targetName has changed', () => {
        const nextProps = {targetName: 'Mars'};
        const minDistance = 3;

        CameraService.getMinDistance = () => minDistance;
        cameraContainer.props = {targetName: 'Earth'};
        cameraContainer.controls.minDistance = 5;
        cameraContainer.maybePreventCameraCollision(nextProps);

        expect(cameraContainer.controls.minDistance).toEqual(minDistance);
      });

      it('should update the minDistance if the scale has changed', () => {
        const targetName = 'Earth';
        const nextProps = {targetName, scale: 5};
        const minDistance = 3;

        CameraService.getMinDistance = () => minDistance;
        cameraContainer.props = {targetName};
        cameraContainer.controls.minDistance = 5;
        cameraContainer.maybePreventCameraCollision(nextProps);

        expect(cameraContainer.controls.minDistance).toEqual(minDistance);
      });
    });

    describe('maybeUpdateAutoOrbit()', () => {
      it('should set controls.autoRotate to the value of `isAutoOrbitEnabled` if it has changed', () => {
        const isAutoOrbitEnabled = false;
        const nextProps = {isAutoOrbitEnabled};
        
        cameraContainer.props = {isAutoOrbitEnabled: !isAutoOrbitEnabled};
        cameraContainer.componentWillReceiveProps(nextProps);

        expect(cameraContainer.controls.autoRotate).toEqual(isAutoOrbitEnabled);
      });

      it('should change the value of controls.autoRotate if `isAutoOrbitEnabled` has not changed', () => {
        const isAutoOrbitEnabled = false;
        const nextProps = {isAutoOrbitEnabled};

        cameraContainer.props = {isAutoOrbitEnabled};
        cameraContainer.componentWillReceiveProps(nextProps);

        expect(cameraContainer.controls.autoRotate).toEqual(isAutoOrbitEnabled);
      });
    });

    describe('maybeUpdateControlsZoom()', () => {
      it('should call controls.zoom() if the `zoom` prop has changed', () => {
        const zoom = 10;
        const nextProps = {zoom: 20};
        const spy = jest.spyOn(cameraContainer.controls, 'zoom');

        cameraContainer.props = {zoom};
        cameraContainer.componentWillReceiveProps(nextProps);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith(nextProps.zoom);
      });

      it('should not call controls.zoom() if the `zoom` prop has not changed', () => {
        const zoom = 10;
        const nextProps = {zoom};
        const spy = jest.spyOn(cameraContainer.controls, 'zoom');

        cameraContainer.props = {zoom};
        cameraContainer.componentWillReceiveProps(nextProps);

        expect(spy).not.toHaveBeenCalled();
      });
    });
  });
 
  describe('cancelTween()', () => {
    beforeEach(() => {
      cameraContainer.tweenBase = {stop: jest.fn()};
    });

    it('should call endTween, if the tweenBase is defined', () => {
      const spy = jest.spyOn(cameraContainer.tweenBase, 'stop');

      cameraContainer.cancelTween();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should dispose of tweenBase', () => {
      cameraContainer.cancelTween();

      expect(cameraContainer).not.toHaveProperty('tweenBase');
    });
  });

  describe('startTween()', () => {
    const pivot = new Object3D();
    const targetName = 'Mars';

    beforeEach(() => {
      cameraContainer.props = {
        scene: {
          add: jest.fn(),
          getObjectByName: jest.fn()
        }
      };
      cameraContainer.refs = {pivot};
      cameraContainer.cancelTween = jest.fn();
      cameraContainer.zoomIn = jest.fn();
      cameraContainer.controls = new Controls(new Camera());

      CameraService.getPivotTween = () => new TWEEN.Tween();
    });

    describe('when the target exists in the scene', () => {
      const target = new Object3D();

      beforeEach(() => {
        cameraContainer.props.scene.getObjectByName = () => target;
      });

      it('should add the pivot to the scene', () => {
        const spy = jest.spyOn(cameraContainer.props.scene, 'add');

        cameraContainer.startTween(targetName);

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith(pivot);
      });

      it('should tween the pivot to the discovered target', () => {
        const spy = jest.spyOn(cameraContainer, 'tweenPivot');

        cameraContainer.startTween(targetName);

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith(target, pivot);
      });
    });

    describe('when the target does not exist', () => {
      it('should not tween', () => {
        const spy = jest.spyOn(cameraContainer, 'tweenPivot');

        cameraContainer.props.scene.getObjectByName = () => null;
        cameraContainer.startTween(targetName);

        expect(spy).not.toHaveBeenCalled();
      });
    });
  });

  describe('tweenPivot()', () => {
    const target = new Vector3();
    const pivot = new Vector3();

    beforeEach(() => {
      cameraContainer.zoomIn = jest.fn();
      CameraService.getPivotTween = () => new TWEEN.Tween();
    });

    it('should assign a new Tween', () => {
      cameraContainer.tweenPivot(target, pivot);

      expect(cameraContainer.tweenBase).toBeInstanceOf(TWEEN.Tween);
    });

    it('should call zoomIn()', () => {
      const spy = jest.spyOn(cameraContainer, 'zoomIn');

      cameraContainer.tweenPivot(target, pivot);

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('zoomIn()', () => {
    it('should call controls.tweenZoom()', () => {
      const changeZoom = jest.fn();
      const tweenZoom = jest.fn();

      cameraContainer.controls = {tweenZoom};
      cameraContainer.props = {action: {changeZoom}};
      
      const spy = jest.spyOn(cameraContainer.controls, 'tweenZoom');

      cameraContainer.zoomIn();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(1, changeZoom);
    });
  });

  describe('update()', () => {
    it('should update the controls', () => {
      cameraContainer.controls = {update: jest.fn()};
      const spy = jest.spyOn(cameraContainer.controls, 'update');

      cameraContainer.update();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('render()', () => {
    it('should render the scene container successfully', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
