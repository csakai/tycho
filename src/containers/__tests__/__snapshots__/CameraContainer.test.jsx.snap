// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera Container render() should render the scene container successfully 1`] = `
<group>
  <perspectiveCamera
    aspect={1.5}
    far={10000}
    fov={50}
    name="camera"
    near={1}
    position={
      Vector3 {
        "add": [Function],
        "addScalar": [Function],
        "addScaledVector": [Function],
        "addVectors": [Function],
        "angleTo": [Function],
        "applyAxisAngle": [Function],
        "applyEuler": [Function],
        "applyMatrix3": [Function],
        "applyMatrix4": [Function],
        "applyProjection": [Function],
        "applyQuaternion": [Function],
        "ceil": [Function],
        "clamp": [Function],
        "clampLength": [Function],
        "clampScalar": [Function],
        "clone": [Function],
        "constructor": [Function],
        "copy": [Function],
        "cross": [Function],
        "crossVectors": [Function],
        "distanceTo": [Function],
        "distanceToManhattan": [Function],
        "distanceToSquared": [Function],
        "divide": [Function],
        "divideScalar": [Function],
        "dot": [Function],
        "equals": [Function],
        "floor": [Function],
        "fromArray": [Function],
        "fromAttribute": [Function],
        "fromBufferAttribute": [Function],
        "getColumnFromMatrix": [Function],
        "getComponent": [Function],
        "getPositionFromMatrix": [Function],
        "getScaleFromMatrix": [Function],
        "isVector3": true,
        "length": [Function],
        "lengthManhattan": [Function],
        "lengthSq": [Function],
        "lerp": [Function],
        "lerpVectors": [Function],
        "max": [Function],
        "min": [Function],
        "multiply": [Function],
        "multiplyScalar": [Function],
        "multiplyVectors": [Function],
        "negate": [Function],
        "normalize": [Function],
        "project": [Function],
        "projectOnPlane": [Function],
        "projectOnVector": [Function],
        "reflect": [Function],
        "round": [Function],
        "roundToZero": [Function],
        "set": [Function],
        "setComponent": [Function],
        "setEulerFromQuaternion": [Function],
        "setEulerFromRotationMatrix": [Function],
        "setFromCylindrical": [Function],
        "setFromMatrixColumn": [Function],
        "setFromMatrixPosition": [Function],
        "setFromMatrixScale": [Function],
        "setFromSpherical": [Function],
        "setLength": [Function],
        "setScalar": [Function],
        "setX": [Function],
        "setY": [Function],
        "setZ": [Function],
        "sub": [Function],
        "subScalar": [Function],
        "subVectors": [Function],
        "toArray": [Function],
        "transformDirection": [Function],
        "unproject": [Function],
        "x": 300,
        "y": 300,
        "z": 300,
      }
    }
  />
</group>
`;
